flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecRank.java

```diff
@@ -0,0 +1,217 @@
+	public static final ConfigOption<Long> TABLE_EXEC_TOPN_CACHE_SIZE =
+		ConfigOptions.key("table.exec.topn.cache-size")
+			.longType()
+			.defaultValue(10000L)
+			.withDescription("TopN operator has a cache which caches partial state contents to reduce" +
+				" state access. Cache size is the number of records in each TopN task.");
+
+	private RankType rankType;
+	private RankProcessStrategy rankStrategy;
+	private RankRange rankRange;
+	private boolean generateUpdateBefore;
+	private boolean outputRankNumber;
+	private int[] partitionFields;
+	private int[] sortFields;
+	private boolean[] sortDirections;
+	private boolean[] nullsIsLast;
+
+	public StreamExecRank(
+			RankType rankType,
+			RankProcessStrategy rankStrategy,
+			RankRange rankRange,
+			boolean generateUpdateBefore,
+			boolean outputRankNumber,
+			int[] partitionFields,
+			int[] sortFields,
+			boolean[] sortDirections,
+			boolean[] nullsIsLast,
+			ExecEdge inputEdge,
+			LogicalType outputType,
+			String description) {
+		super(Collections.singletonList(inputEdge), outputType, description);
+		this.rankType = rankType;
+		this.rankStrategy = rankStrategy;
+		this.rankRange = rankRange;
+		this.generateUpdateBefore = generateUpdateBefore;
+		this.outputRankNumber = outputRankNumber;
+		this.partitionFields = partitionFields;
+		this.sortFields = sortFields;
+		this.sortDirections = sortDirections;
+		this.nullsIsLast = nullsIsLast;
+	}
+
+	@Override
+	protected Transformation<RowData> translateToPlanInternal(PlannerBase planner) {
+		switch (rankType) {
+			case ROW_NUMBER:
+				break;
+			case RANK:
+				throw new TableException("RANK() on streaming table is not supported currently");
+			case DENSE_RANK:
+				throw new TableException("DENSE_RANK() on streaming table is not supported currently");
+			default:
+				throw new TableException(String.format("Streaming tables do not support %s rank function.", rankType));
+		}
+
+		TableConfig tableConfig = planner.getTableConfig();
+
+		RowType inputType = (RowType) getInputNodes().get(0).getOutputType();
+		InternalTypeInfo<RowData> inputRowTypeInfo = InternalTypeInfo.of(inputType);
+		RowDataKeySelector sortKeySelector = KeySelectorUtil.getRowDataSelector(sortFields, inputRowTypeInfo);
+		LogicalType[] sortKeyTypes = IntStream.of(sortFields).mapToObj(inputType::getTypeAt).toArray(LogicalType[]::new);
+		int[] sortKeyPositions = IntStream.range(0, sortFields.length).toArray();
+		GeneratedRecordComparator sortKeyComparator = ComparatorCodeGenerator.gen(
+				tableConfig,
+				"StreamExecSortComparator",
+				sortKeyPositions,
+				sortKeyTypes,
+				sortDirections,
+				nullsIsLast);
+		long cacheSize = tableConfig.getConfiguration().getLong(TABLE_EXEC_TOPN_CACHE_SIZE);
+		long minIdleStateRetentionTime = tableConfig.getMinIdleStateRetentionTime();
+		long maxIdleStateRetentionTime = tableConfig.getMaxIdleStateRetentionTime();
+
+		AbstractTopNFunction processFunction;
+		if (rankStrategy instanceof AppendFastStrategy$) {
```

Comment:
it's better we could port `AppendFastStrategy` to Java ? it's a little strange to see `$`